explainer/effect-generic-bounds-and-functions.md

```diff
@@ -0,0 +1,544 @@
+- Feature Name: `effect-generic-bounds-and-functions`
+- Start Date: (2024-01-20)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+[RFC 0000](./effect-generic-trait-declarations.md) introduces traits which are
+generic over an effect, but implementers have to pick whether they want to
+implement the base version or the effectful version of the trait. This RFC
+extends that system further by removing that limitation, and enabling authors to
+write functions which themselves are generic over effects. For example, here is
+a function `io::copy` which would be able to operate either synchronously or
+asynchronously, depending on which types are passed to it.
+
+```rust
+/// This defines a trait `Read` which may or may not
+/// be async, using the design introduced in RFC 0000.
+#[maybe(async)]
+pub trait Read {
+    #[maybe(async)]
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
+}
+
+/// This defines a trait `Write` which may or may not
+/// be async, using the design introduced in RFC 0000.
+#[maybe(async)]
+pub trait Write {
+    #[maybe(async)]
+    fn write(&mut self, buf: &[u8]) -> Result<usize>;
+    #[maybe(async)]
+    fn flush(&mut self) -> Result<()>;
+}
+
+/// This defines a function `copy` which copies bytes from a
+/// reader into a writer. This RFC enables this function to
+/// operate either synchronously or asynchronously. Where if
+/// operating synchronously, the `.await` operator becomes a
+/// no-op.
+#[maybe(async)] 
+pub fn copy<R, W>(reader: &mut R, writer: &mut W) -> Result<()>
+where
+    R: #[maybe(async)] Read + ?Sized,
+    W: #[maybe(async)] Write + ?Sized,
```

Comment:
This isn't something to fix in the draft, but mostly just a comment: this kind of further cements the fact that the placeholder attribute syntax isn't ideal, since it's not 100% clear that the maybes are *all* linked, and that you can only have async-all or sync-all, and not a mixed async and sync.

Reply:
Yeah fair point; that was also an issue brought up in #45. We probably need to find a way to better address that.